### PR TITLE
V2.0.0

### LINKED ISSUE #87: Add feedbacks based on the sources of an inputs layers/multivew sources
**Describe the feature**
Add feedback based on the current input being used as a specific layer in an(other) inputs multiview/layers.

**Is this version dependent (vMix v24.)?**
No

**If documentation is required to implement, do you know where to find it?**
If applicable, add screenshots, protocol description PDFs, etc., to help the developers along.

**Usecases**
In concert with the ability to set the source for an inputs various layers, it would be useful to be able to display feedback based on the layers. e.g. Make background green when input 12 is set as layer 3 on input 2

### LINKED ISSUE #90: RangeError: Invalid string length
**Describe the bug**
Sentry is reporting a lot of errors with the line
https://github.com/bitfocus/companion-module-studiocoast-vmix/blob/master/src/tcp.js#L63

`RangeError: Invalid string length`

57k occurances from 6 users in roughly the last month.

**System (please complete the following information):**
 - Companion Version: multiple including 2.2.0-3e5a8eb-2868
 - vMix version [e.g. v23]

### LINKED ISSUE #105: Layer Tally indicators don't work with new(ish) companion feature that removes the top bars from buttons
The layer tally indicator feedback system seems to be broken if you choose to hide all the top bars from buttons. With that option un-ticked the indicators work as I anticipated, with it ticked I get no useful feedback, and a strange holdover of full button feedback if I go from a simple input (feedback if previewed for example) to another input that uses the first one on a layer. This is hard to explain in words!

### LINKED ISSUE #111: X/Y Layers / Multiview Name & GUID not working
**Describe the bug**

Some Field do not work with Name & GUID, (even if its indicated it should), They do work with Input# though.

Layers / MultiView - Set Destination Input for Routable Multiview Overlay Layer
> Destination MV Input

Layers / MultiView - check selected destination input
> Input (Nr/Name/GUID) 

Screenshot 2021-07-13 094314

**Expected behaviour**
to work with Input#, Name & GUID

**System (please complete the following information):**

- Companion Version: companion-2.2.0-92a53337-3474-win64
- vMix version: 24.0.0.63

**Additional context**
Forced to use vMix Dynamic's instead of this nice module, Input# are just not reliable enough since they can be easily changed just by a miss click...

### LINKED ISSUE #117: Feedback Overlay-Overlay status : Foreground Color issue
**Feedback Overlay-Overlay status**
Changes in this function is creating Foreground color not to work correctly.

The Foreground color is now linked between the Preview and Program. And it changes the Foreground color if its either on Preview or Program when like in the attached image you use 1 button for preview and another one for program.

BadOverlayColor

To fix this maybe have a drop down to choose what the feedback will be affected by: Preview / Program / Preview+Program like I have seen on other feedback functions.

Companion Version: companion-2.2.0-92a53337-3474-win64
vMix version: 24.0.0.63

### LINKED ISSUE #119: Allow for DynamicX or Custom Variables to be used in vMix custom commands
**Description of feature request**
When sending commands from Companion to vMix the variables _Dynamic1_ through _Dynamic4_ can be used to affect specific layers/values/texts without the need to create duplicate sets of buttons. However: This is not possible when sending custom commands in plain text using _vMix: Run custom command_
The same is true for the Companion-internal custom cariables e.g. $(internal:custom_variable_name)

**Is this version dependent (vMix v24.)?**
Using _vMix: Run custom command_: Not dependant on version.
Using vMix's _DynamicX_ : requires a version with cynamic variables implemented.

**If documentation is required to implement, do you know where to find it?**
If it is implemented on the Companion end, the feature should be "easy" to add. However, I am not certain the vMix API can handle a request should the solution rely on _DynamicX_.

**Usecases**
I have an overlay graphic in Vmix with multiple graphical elements. It is a scoreboard for team sports and I want to show/hide different image elements (using _SetImageVisible_) for the home team as well as the guests. 

Currently I am restricted to the two following solutions:
- Split the elements into standalone Inputs, which is more difficult to line up properly.
- Create duplicate (or more) buttons in Companion which takes up more real easte.

**Sidenote**
An entirely different request would be to implement _SetImageVisible_ as a function similar to _vMix: vmix: Title - Adjust title text_  where there is an option to pass _SelectedIndex_ as a variable in an input field.

### LINKED ISSUE #129: TypeError: Cannot read property 'shortTitle' of undefined
**Describe the bug**
Sentry is reporting a lot of errors with the line:
https://github.com/bitfocus/companion-module-studiocoast-vmix/blob/6e96cf41e83298c8abb39efbb0a0f16ced31d1b1/src/api.js#L561

Additionally, a similar error is being thrown from: https://github.com/bitfocus/companion-module-studiocoast-vmix/blob/6e96cf41e83298c8abb39efbb0a0f16ced31d1b1/src/api.js#L349 and https://github.com/bitfocus/companion-module-studiocoast-vmix/blob/6e96cf41e83298c8abb39efbb0a0f16ced31d1b1/src/api.js#L362

48k+ occurances across 75 users in the last 30 days

- Companion Version: Various 2.2.0 beta builds, including build 3711 (11th jan 2022)
- vMix version [e.g. v23]

**Additional context**

### LINKED ISSUE #135: Trigger Variable for vMix Tally Status of an input
It would be great to surface the tally status of any given input (by GUID or ID) as a trigger variable.

Right now there is a mix_1_program (and preview) variable, but only if that input is in PGM, and doesn't take into account if a source is in a layer, overlay or multiview.  In short it would be helpful to have option for functionality mirror the native vMix tally state.

Example Use Case:

I'm using this as a way of integrating tally lights from vMix with other tally systems.  I thought having control in Companion with new Triggers may be a way for people that are familiar with Companion to implement easy tally systems.  I wrote a vMix to  Blackmagic 3G SDI Arduino implementation that I thought might be even better to use Companion as the tally trigger middleware.

### LINKED ISSUE #136: More companion variables
**Describe the feature**
The vmix API gives a lot more information about the inputs that can be usefull with the new trigger function. The type of the input, for example. 
Trigger example: $(vmix:mix_1_program_type) = video run the command to play the video.

**Is this version dependent (vMix v24.)?**
No

**Usecases**
Add a couple of use cases for why this should be implemented and when/why to use it.

### LINKED ISSUE #140: Connect via hostname also instead of just IP
**Describe the feature**
Connect via hostname also instead of just IP address

I am not sure of what this breaks, but I would think it would "just work" ????

**Usecases**
Deployment of our systems would be easier

### PR DESCRIPTION
**Pull request by thedist**
Rewrite in TypeScript
Significant overhaul of instance variables
standardized actions/feedback/variables structure
3 TCP sockets to split traffic for XML data, Activators, and Functions
Better TCP message handling
Added vMix v25 Actions, Feedback, and Activator handlers
Added button shifting feature
Loads more Instance Variables
Deprecated titleLayer, inputSelectedIndexName, and multiviewLayer Feedbacks
Deprecated input_volume_${input.number} and input_volume_${inputName} instance variables

Resolves #87
Resolves #90
Resolves #105
Resolves #111
Resolves #117
Resolves #119
Resolves #129
Resolves #135 
Resolves #136 
Resolves #140 

This is a big update, and while I've done a lot of testing myself, and with a few others (end users, not developers) I had to rush a little recently to get it ready for Companion 2.2.0 deadline, and would appreciate some people testing.

This update DOES have an upgrade script, and there are breaking changes, so make sure you BACKUP YOUR CONFIGURATION before running, as it's not possible to reverse.

### CODE REVIEW

File: src/variables.ts [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+      { label: 'Dynamic Value 3', name: 'dynamic_value_3' },
+      { label: 'Dynamic Value 4', name: 'dynamic_value_4' },
+
+      // Layers
+      { label: 'Layer Routing Input', name: 'layer_routing_input' },
+      { label: 'Layer Routing Layer', name: 'layer_routing_layer' },
+    ])
+
+    // Inputs
+    const inputNumberVariables: Set<InstanceVariableDefinition> = new Set([])
```
Comment (Julusian) [inline src/variables.ts:193]: You can also write this as `const inputNumberVariables = new Set<InstanceVariableDefinition>()`

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,36 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "es2019",
+    "strict": true,
+    "alwaysStrict": false,
+    "strictPropertyInitialization": true,
```
Comment (Julusian) [inline tsconfig.json:7]: a bunch of these options fall under the `strict` flag, so you dont have to explicitly enable them. but there is no harm in doing so

**Review by Julusian — APPROVED** [review]
I havent spotted anything bad here, I only have some small suggestions and comments.

As you have unit tests, it might be good to setup running those in github actions in this repository?

File: package.json [hunk 3]
```diff
@@ -10,18 +10,64 @@
 	"product": "vMix",
 	"shortname": "vmix",
 	"description": "Module to control StudioCoast vMix",
-	"main": "src/index.js",
+	"main": "dist/index.js",
 	"scripts": {
-		"test": "echo \"Error: no test specified\" && exit 1"
+		"dev": "yarn build:watch",
+		"build": "npx rimraf dist && npx typescript@^4.2.3 -p tsconfig.build.json",
```
[the remark this answers is not in the document]
Comment (thedist) [reply package.json]: Thanks. I had updated all of the dependencies before submitting the PR but forgot to update the version in the script.

**Comment by thedist** [comment]
I've resolved those changes suggested by Julusian,.

Using GitHub Actions for CI testing I'll add later on as I'm not familiar enough with them currently. Most of my work utilizes AWS Code Pipeline to handle that side of things for my own infrastructure so I haven't had to touch GitHub Actions yet.

As for the boolean feedback, I'm a bit on the fence about that as a significant portion of feedback requires the advanced type, and if we was to use boolean feedbacks and then need to switch to advanced in the future then it would be a breaking change to anyone using them as triggers. With the significant increase in Instance Variables I'm hoping that they will fill the need for triggers as they are quite comprehensive now.

I'm glad you liked these types being a bit stricter than the Companion types. While it has been a pain to try and learn TypeScript and being strict from the start, I feel being less permissive will lead to fewer errors as a lot of the issues this module has had in the past has been due to options with an undefined value, simple typos, inconsistent casing etc... I hope to do more in the future, not just with this module but with Companions itself (for example the TCP module I'm importing from Companion has no types, I plan to add types to that at some point if no one else has plans to).

**Comment by Julusian** [comment]
> I've resolved those changes suggested by Julusian,.
> 
> Using GitHub Actions for CI testing I'll add later on as I'm not familiar enough with them currently. Most of my work utilizes AWS Code Pipeline to handle that side of things for my own infrastructure so I haven't had to touch GitHub Actions yet.

yeah, its a little fiddly because of the typings inherited from companion.
You could either use this as a starting point https://github.com/bitfocus/companion-module-youtube-live/blob/master/.github/workflows/postcommit.yml
or https://github.com/bitfocus/companion/blob/master/.github/workflows/test.yml and manually wget the needed d.ts files into the right place first. 

> As for the boolean feedback, I'm a bit on the fence about that as a significant portion of feedback requires the advanced type, and if we was to use boolean feedbacks and then need to switch to advanced in the future then it would be a breaking change to anyone using them as triggers. With the significant increase in Instance Variables I'm hoping that they will fill the need for triggers as they are quite comprehensive now.

True, but do you need to make the feedbacks more complex? I much prefer having smaller more granular feedbacks. 
for example, the 'Tally - Overlay state' feedback could easily be configured by a pair of feedbacks, one which checks if it is on program and one checking for on preview. With the bonus of the times I dont care about it being on preview I can have it not do that one
And by doing that as a pair of boolean feedbacks, as a user I can decide that I dont want to change the background colour each time, but instead I want to use a different png. 
Which is the other key thing with boolean feedbacks. They also let the user choose what style properties they want to change.

> I'm glad you liked these types being a bit stricter than the Companion types. While it has been a pain to try and learn TypeScript and being strict from the start, I feel being less permissive will lead to fewer errors as a lot of the issues this module has had in the past has been due to options with an undefined value, simple typos, inconsistent casing etc... I hope to do more in the future, not just with this module but with Companions itself (for example the TCP module I'm importing from Companion has no types, I plan to add types to that at some point if no one else has plans to).

I beat you to it, seeing the import while reading this pr made me want to have some types written :) https://github.com/bitfocus/companion/blob/master/tcp.d.ts

**Pull request merged**